From 827ca0e97936adc8a58fe056d663156671f713c5 Mon Sep 17 00:00:00 2001
From: Florian Schmidt <florian.schmidt.stargatewissen@gmail.com>
Date: Tue, 19 Dec 2017 17:22:16 +0100
Subject: [PATCH] [mergeMessageFileList] Change path to $IP when using extension-dir too

Imagine a setup of mediawiki is deployed from another host, which has
a different "basepath" to MediaWiki (e.g. something prepended with deployment
or whatever). The setup may use the mergeMessageFileList.php maintenance script
to create a merged version of all messages files for using the
rebuildLocalisation.php maintenance script to create the i18n cache files. This
all happens on the deployment host (with the prepended deployment/ path).

However the extensions could be symlinked on the deployment host, only, and
are not located in ../extensions relatively from the maintenance script directory,
which means, that everything is working when using the extensions-dir parameter to
the mergeMessageFileList.php maintenance script. But the output differs from the
expected one. Instead of having paths relative to $IP, the paths in these files
are absolute (e.g. /data/deployment/extensions/ExtName, instead of $IP/ExtName).
This is, because only some limited variables are replaced with $IP after the script
is finished, excluding the paths added by extensions-dir, which are of interest in
this run.

This change adds all paths passed as extensions-dir to the script to the list of
paths that will be replaced by $IP after mergeMessageFileList finished.

Change-Id: Ic26dffb58366b608add01496287bce32258bdafe
---

diff --git a/maintenance/mergeMessageFileList.php b/maintenance/mergeMessageFileList.php
index 48a6666..fab97a5 100644
--- a/maintenance/mergeMessageFileList.php
+++ b/maintenance/mergeMessageFileList.php
@@ -28,6 +28,9 @@
 require_once __DIR__ . '/Maintenance.php';
 $maintClass = MergeMessageFileList::class;
 $mmfl = false;
+$toReplaceIPDirs = [
+	dirname( __DIR__ ),
+];
 
 /**
  * Maintenance script that merges $wgExtensionMessagesFiles from various
@@ -53,7 +56,7 @@
 
 	public function execute() {
 		// phpcs:ignore MediaWiki.NamingConventions.ValidGlobalName.wgPrefix
-		global $mmfl;
+		global $mmfl, $toReplaceIPDirs;
 		global $wgExtensionEntryPointListFiles;
 
 		if ( !count( $wgExtensionEntryPointListFiles )
@@ -78,6 +81,9 @@
 			# Allow multiple directories to be passed with ":" as delimiter
 			$extdirs = explode( ':', $extdir );
 			foreach ( $extdirs as $extdir ) {
+				// strip extensions/ or skins/ from the directory name
+				$toReplaceIPDirs[] = dirname( $extdir );
+
 				$entries = scandir( $extdir );
 				foreach ( $entries as $extname ) {
 					if ( $extname == '.' || $extname == '..' || !is_dir( "$extdir/$extname" ) ) {
@@ -191,13 +197,9 @@
 	'$wgExtensionMessagesFiles = ' . var_export( $wgExtensionMessagesFiles, true ) . ";\n\n" .
 	'$wgMessagesDirs = ' . var_export( $wgMessagesDirs, true ) . ";\n\n";
 
-$dirs = [
-	$IP,
-	dirname( __DIR__ ),
-	realpath( $IP )
-];
-
-foreach ( $dirs as $dir ) {
+$toReplaceIPDirs[] = $IP;
+$toReplaceIPDirs[] = realpath( $IP );
+foreach ( $toReplaceIPDirs as $dir ) {
 	$s = preg_replace( "/'" . preg_quote( $dir, '/' ) . "([^']*)'/", '"$IP\1"', $s );
 }
 
